### ISSUE TITLE
Array to Hash snippet

### ISSUE THREAD

**Issue by birula** [codereading/rack#3]
I was looking around the code and found this piece of code that turns an array into a Hash. The basic idea is to flatten an array to create a list of key, value and then pass it into a Hash[].

https://github.com/codereading/rack/blob/rack-1.4/lib/rack/request.rb#L57

``` ruby
ary = [ [:foo, "bar"], [:fii, "baz"] ].flatten    #=> [ :foo, "bar", :fii, "baz"]
Hash[ary]                                         #=> {"foo"=> "bar", "fii"=> "baz"}

Hash[*content_type.split(/\s*[;,]\s*/)[1..-1].
  collect { |s| s.split('=', 2) }.
  map { |k,v| [k.downcase, v] }.flatten]
```

This may be not a big thing but I brought it on because I've used it a couple of times.

**Comment by Monomachus** [comment]
what does this other line?

Hash[_content_type.split(/\s_[;,]\s*/)[1..-1].
  collect { |s| s.split('=', 2) }.
  map { |k,v| [k.downcase, v] }.flatten]

**Comment by samnang** [comment]
@birula I think you are missing splat operator `Hash[ary]` in your example, so what it should be is `Hash[*ary]`.

Another interesting `collect` or `map` is just an alias of another. Why don't they keep using one consistant instead of mixing in using all together?

**Comment by birula** [comment]
@samnang You're right i forgot \* before the array, just don't now why is it used.

Doesn't it sound like you're trying to map a [key, value] pairs and collect anything else? It looks silly but i use those methods that way.

**Comment by birula** [comment]
@Monomachus this line convert the header information from a request "text/html; charset=utf-8" into an array [["charset", "utf-8"]] and then to a hash { "charset" => "utf-8" }

If you look closely it excludes the "text/html" from the string in the first step.

**Comment by birula** [comment]
There's another interesting thing in this code, after the first split it gets the last element with [1..-1] why not just use the .last method ?

``` ruby
content_type.split(/\s*[;,]\s*/)[1..-1]
content_type.split(/\s*[;,]\s*/).last
```

**Comment by tomykaira** [comment]
@birula, try more than one media parameters.

`last` takes only the last element, whereas `[1..-1]` takes other than the first element.
Plus, the former returns an object in array, the latter returns an array.

**Comment by birula** [comment]
@tomykaira you're completely right using last in this case is wrong.

**Comment by razum2um** [comment]
would it be more efficiently? (or not?)
i.e. would [1..-1] allocate one more array?

``` ruby
content_type.split(/\s*[;,]\s*/).tap(&:shift)
```
